Clean up ChainSetup and setup helpers tests

hobofan opened the pull request:
Fixes #23.

This sets up a integration test environment similar to the one from `mosaic-contracts`.
Changes made compared to the old tests (previously in `tests/helpers/setup`):
- General code cleanup
- The tests now use two seperate geth nodes that are started during integration test setup
- Deployed contracts are reused between integration test steps
- Config is now provided via the `00_config` integration test step and not pulled from a directory outside of the repository anymore

Additionally to converting the old tests, I added some simple test for the activation of a deployment (setting coAnchor, setting CoGateway on OSTPrime and activating Gateway).

This PR doesn't change anything about the code in `libs` (apart from some default gas value reductions), but should provide a good test coverage before refactoring.

hobofan committed:
Clean up SetupConfig tests and setup integration tests

hobofan committed:
Clean up AnchorHelper tests

hobofan committed:
Clean up LibsHelper tests

hobofan committed:
Clean up OrganizationHelper tests

hobofan committed:
Clean up OSTPrimeHelper tests

hobofan committed:
Clean up GatewayHelper tests

hobofan committed:
Clean up CoGatewayHelper tests

hobofan committed:
add license headers

hobofan committed:
move tests of individual setup steps into 02_sequential_setup

hobofan committed:
correctly reuse deployed contracts between integration tests

hobofan committed:
add integration tests for deployment activation

schemar commented on test_integration/01_chain_setup/01_chain_setup.js:
```diff
@@ -0,0 +1,113 @@
+const MockContractsDeployer = require('../../tests/utils/MockContractsDeployer');
+
+describe('ChainSetup', () => {
+  let caMockToken;
```
What is `ca`? 🤔

hobofan replied:
ca is contract address. I've usually cleaned it up to address, but looks like I missed this one.

schemar commented on test_integration/01_chain_setup/01_chain_setup.js:
```diff
@@ -0,0 +1,113 @@
+      }
+    };
+
+    return helper.setup(valueToken, originConfig, auxiliaryConfig).then((output) => {
```
Why isn't `chainSetup` a function instead of a class? I don't see the usefulness of the `helper` variable here, but maybe it is used in other contexts.
Another alternative would be for example to call the variable `setup` and the method `run`.

hobofan replied:
Changing the interface of ChainSetup is out of scope of this ticket

schemar replied:
👍

schemar commented on test_integration/02_sequential_setup/01_libs_helper.js:
```diff
@@ -0,0 +1,147 @@
+  let addressMessageBus;
+  let addressGatewayLib;
+
+  let helper = new LibsHelper(shared.origin.web3, addressMerklePatriciaProof, addressMessageBus, addressGatewayLib);
```
If you call the variable `helper` here it is super ambiguous further down as we have dozens of things called `...Helper` 😳

hobofan replied:
In these tests `helper` is supposed to be the `subject` of the tests. I'll rename it to that.

schemar commented on test_integration/02_sequential_setup/01_libs_helper.js:
```diff
@@ -0,0 +1,147 @@
+      from: shared.setupConfig.deployerAddress,
+      gasPrice: shared.setupConfig.gasPrice
+    };
+    return helper
```
What does it mean if an `it` block returns?

hobofan replied:
The returned value is a Promise. Alternatively we could more strictly use async/await.

schemar replied:
And then Mocha waits for it to resolve?

hobofan replied:
Yes. If you are creating promises and don't return them at the end, Mocha doesn't know about them and assumes that the test case has finished. Then it isn't able to correctly handle assert failures (and might ignore them).

schemar replied:
Ah, ok good to know about the return 👍 Thanks!

schemar commented on test_integration/02_sequential_setup/02_organization_helper.js:
```diff
@@ -0,0 +1,118 @@
+
+  it('should set admin address', () => {
+    //Admin Key Address
+    let kaAdmin = orgConfig.admin;
```
What is `ka`? Admin key address?
Then why not write `adminKeyAddress` and remove the comment above? Makes reading much easier further down when I don't need to go back up to check what `ka` meant.
Same for other cases and possibly `ca`. Does it mean `contractAddress`?? Then why not write it??

hobofan replied:
👍

schemar commented on test_integration/02_sequential_setup/03_anchor_helper.js:
```diff
@@ -0,0 +1,143 @@
+      });
+  });
+
+  // Set Co-Anchor Address.
```
Superfluous comment (also in other places).

hobofan replied:
👍

schemar reviewed (changes requested):
Looks like good coverage 👍 

Calling everything `helper` is not useful in my opinion. It even decreases readability if you ask me. I would even argue that use of `helpers` often indicates an architectural issue which can be solved. Especially excessive use.

It seems a lot of `let`s should be exchanged for `const`s.

I think all comments should start with an uppercase letter and end with a period (English sentences).

How does this relate to #33? It seems some things exist in both?

Also, maybe not all my comments apply to the scope of this ticket?

deepesh-kn commented on test_integration/01_chain_setup/01_chain_setup.js:
```diff
@@ -0,0 +1,113 @@
+
+'use strict';
+
+const { assert } = require('chai');
```
assert is not used in the code.

schemar replied:
🦅👁

deepesh-kn commented on test_integration/01_chain_setup/01_chain_setup.js:
```diff
@@ -0,0 +1,113 @@
+'use strict';
+
+const { assert } = require('chai');
+const Web3 = require('web3');
```
`Web3` is not used in the file.

deepesh-kn commented on test_integration/02_sequential_setup/01_libs_helper.js:
```diff
@@ -0,0 +1,147 @@
+'use strict';
+
+const { assert } = require('chai');
+const Web3 = require('web3');
```
`assert` and `Web3` is not used in file.

hobofan replied:
In this file they are ;)

deepesh-kn reviewed (commented):
I do not have major comments.
For me the tests are not running with `npm run test:integration`
```
npm install
npm run test:integration

> @openstfoundation/mosaic.js@0.1.0 test:integration /Users/deepeshkn/Documents/workspace/hobofan/mosaic.js
> node test_integration/integration_tests.js


```
Nothing happens after this.

schemar reviewed (approved):
After understanding the scope and prerequisites of the ticket, I think this PR is very good and basically mergeable. If you can make some improvements before, then that would be great 👍

hobofan commented:
@deepesh-kn Did it maybe just take some time to pull the docker images (which makes the first run take longer)? I'll add a log line that warns about that.

deepesh-kn commented:
> @deepesh-kn Did it maybe just take some time to pull the docker images (which makes the first run take longer)? I'll add a log line that warns about that.

I did not kill it since it was run and still its stuck. I don't even know if it's running in the background (but its long time now).

hobofan committed:
Address PR feedback

- hook up integration tests in Travis
- helper -> subject
- let -> const where possible
- "ka" -> "keyAddress"
- remove useless parameters supplied to *Helper constructors in tests

hobofan committed:
Improve handling of unavailable docker daemon

hobofan commented:
I had a call with @deepesh-kn, end the problem ended up being that the docker daemon was not running and the error was swallowed by the test runner. This case is now handled more gracefully by the last commit.
